packages/main/src/plugin/kubernetes/contexts-manager-experimental.ts

```diff
@@ -16,12 +16,54 @@
  * SPDX-License-Identifier: Apache-2.0
  ***********************************************************************/
 
-import type { KubeConfig, KubernetesObject } from '@kubernetes/client-node';
+import type { Context, KubernetesObject } from '@kubernetes/client-node';
+import { KubeConfig } from '@kubernetes/client-node';
 
 import type { ContextGeneralState, ResourceName } from '/@api/kubernetes-contexts-states.js';
 
+import { ContextsConnectivityRegistry } from './contexts-connectivity-registry.js';
+import { HealthChecker } from './health-checker.js';
+
+const HEALTH_CHECK_TIMEOUT_MS = 5_000;
+
 export class ContextsManagerExperimental {
-  async update(_kubeconfig: KubeConfig): Promise<void> {}
+  // kubeConfigCheck is used by `update` to check if the received kubeconfig has changed
+  // its value must not be used except for detecting this change
+  // Methods of this class should receive a kubeconfig from `update`
+  #kubeConfigCheck = JSON.stringify(new KubeConfig());
+
+  // Registry to store connectivity state of contexts
+  #contextsConnectivityRegistry = new ContextsConnectivityRegistry();
+
+  // Registry to store health checkers, to be able to abort them when a new update is called
+  #healthCheckers = new Map<string, HealthChecker>();
+
+  async update(kubeconfig: KubeConfig): Promise<void> {
+    const kubeConfigToCheck = JSON.stringify(kubeconfig);
+    if (kubeConfigToCheck === this.#kubeConfigCheck) {
+      // the config did not change since last update, do nothing
+      return;
+    }
+
+    this.disposeAllHealthChecks();
+
+    this.#kubeConfigCheck = kubeConfigToCheck;
+
+    for (const kubeContext of kubeconfig.getContexts()) {
+      const contextName = kubeContext.name;
+      const isolatedConfig = this.getIsolatedKubeConfig(kubeconfig, kubeContext);
+      const healthChecker = new HealthChecker(isolatedConfig);
+      this.#healthCheckers.set(contextName, healthChecker);
+      healthChecker.onReadiness((ready: boolean) => {
+        this.#contextsConnectivityRegistry.setChecking(contextName, false);
+        this.#contextsConnectivityRegistry.setReachable(contextName, ready);
+      });
+      this.#contextsConnectivityRegistry.setChecking(contextName, true);
+      healthChecker
+        .checkReadiness({ timeout: HEALTH_CHECK_TIMEOUT_MS })
+        .catch((err: unknown) => console.error(`error checking readiness for context ${contextName}: ${err}`));
+    }
```

Comment:
IMHO it's very complex to read.

I don't see why there is a checking thing in a separate 3rd party registry

AFAIK, if there is a health checker object attached, it means that this context is being monitored.